qutip/qip/circuit_latex.py

```diff
@@ -42,31 +47,121 @@
+    with open(file_stem + ".pdf", "rb") as file:
+        return file.read()
+
 
-    with open(filename + ".tex", "w") as file:
-        file.write(_latex_template % (_qcircuit_latex_min, code))
+# Record type to hold definitions of possible conversions - this is just for
+# reading convenience.
+_PossibleConverter = collections.namedtuple(
+    '_PossibleConverter',
+    ['file_type', 'dependency', 'executables', 'arguments', 'binary'],
+)
+CONVERTERS = {"pdf": _convert_pdf}
+_POSSIBLE_CONVERTERS = [
+    _PossibleConverter('png', 'ImageMagick', ['magick', 'convert'],
+                       arguments=('-density', '100'), binary=True),
+    _PossibleConverter('svg', 'pdf2svg', ['pdf2svg'],
+                       arguments=(), binary=False),
+]
 
-    os.system("pdflatex -interaction batchmode %s.tex" % filename)
-    os.system("rm -f %s.aux %s.log" % (filename, filename))
-    os.system("pdfcrop %s.pdf %s-tmp.pdf" % (filename, filename))
-    os.system("mv %s-tmp.pdf %s.pdf" % (filename, filename))
 
-    if format == 'png':
-        os.system("convert -density %s %s.pdf %s.png" % (100, filename,
-                                                         filename))
-        with open("%s.png" % filename, "rb") as f:
-            result = f.read()
+def _make_converter(possible):
+    """
+    Create the actual conversion function of signature
+        file_stem: str -> 'T,
+    where 'T is data in the format to be converted to.
+    """
+    which = _find_system_command(possible.executables)
+    if which is None:
+        return None
+    mode = "rb" if possible.binary else "r"
+
+    def converter(file_stem):
+        in_file = file_stem + ".pdf"
+        out_file = file_stem + "." + possible.file_type
+        _run_command((which, *possible.arguments, in_file, out_file))
+        with open(out_file, mode) as file:
+            return file.read()
+    return converter
+
+
+for possible in _POSSIBLE_CONVERTERS:
+    # Make the converter using a higher-order function, because if we defined a
+    # function in the loop, it would be easy to later introduce bugs due to
+    # leaky closures over loop variables.
+    converter = _make_converter(possible)
+    if converter:
+        CONVERTERS[possible.file_type] = converter
     else:
-        os.system("pdf2svg %s.pdf %s.svg" % (filename, filename))
-        with open("%s.svg" % filename) as f:
-            result = f.read()
+        message = "".join(["Could not find system ", possible.dependency, ".",
+                           " Image conversion to '", possible.file_type, "'",
+                           " is unavailable."])
+        # ImportWarning is ignored by default, so the user won't actually see
+        # this warning...
+        warnings.warn(message, category=ImportWarning)
+
 
-    return result
+if not _pdflatex:
+    def make_image(*args, **kwargs):
+        raise NotImplementedError("Could not find system 'pdflatex'.")
```

Comment:
Is this a `NotImplementedError`? An external dependency is missing, sounds to me more like a more general `RuntimeError`?

Reply:
Yeah, you're probably right on that one.